common/src/main/java/com/microsoft/identity/common/internal/logging/Logger.java

```diff
@@ -0,0 +1,248 @@
+    private static final Logger INSTANCE = new Logger();
+    private static final String DATE_FORMAT = "yyyy-MM-dd HH:mm:ss";
+
+    // Turn on the VERBOSE level logging by default.
+    private LogLevel mLogLevel  = LogLevel.VERBOSE;
+    private AtomicReference<ILoggerCallback> mExternalLogger = new AtomicReference<>(null);
+
+    // Disable to log PII by default.
+    private static boolean mAllowPii = false;
+    // Disable to Logcat logging by default.
+    private static boolean mAllowLogcat = false;
+
+    /**
+     * Enum class for LogLevel that the sdk recognizes.
+     */
+    public enum LogLevel {
+        /**
+         * Error level logging.
+         */
+        ERROR,
+        /**
+         * Warn level logging.
+         */
+        WARN,
+        /**
+         * Info level logging.
+         */
+        INFO,
+        /**
+         * Verbose level logging.
+         */
+        VERBOSE
+    }
+
+    /**
+     * @return The single instance of {@link Logger}
+     */
+    public static Logger getInstance() {
+        return INSTANCE;
+    }
+
+    /**
+     *  Enable/Disable log message with PII (personal identifiable information) info.
+     *  By default, the SDK doesn't log any PII.
+     *
+     * @param allowPii True if enabling PII info to be logged, false otherwise.
+     */
+    public static void setAllowPii(final boolean allowPii) {
+        mAllowPii = allowPii;
+    }
+
+    /**
+     * Enable/Disable the Android logcat logging. By default, the sdk disables it.
+     *
+     * @param allowLogcat True if enabling the logcat logging, false otherwise.
+     */
+    public static void setAllowLogcat(final boolean allowLogcat) {
+        mAllowLogcat = allowLogcat;
+    }
+
+    /**
+     * Get if log PII is enabled.
+     */
+    public static boolean getAllowPii() {
+        return mAllowPii;
+    }
+
+    /**
+     * Get if logcat is enabled.
+     */
+    public static boolean getAllowLogcat() {
+        return mAllowLogcat;
+    }
+
+    /**
+     * Set the log level for diagnostic purpose. By default, the sdk enables the verbose level logging.
+     *
+     * @param logLevel The {@link LogLevel} to be enabled for the diagnostic logging.
+     */
+    public void setLogLevel(final LogLevel logLevel) {
+        mLogLevel = logLevel;
+    }
+
+    /**
+     * Set the custom logger. Configures external logging to configure a callback that
+     * the sdk will use to pass each log message. Overriding the logger callback is not allowed.
+     *
+     * @param externalLogger The reference to the {@link ILoggerCallback} that can
+     *                       output the logs to the designated places.
+     * @throws IllegalStateException if external logger is already set, and the caller is trying to set it again.
+     */
+    public void setExternalLogger(final ILoggerCallback externalLogger) {
+        if (externalLogger == null) {
+            return;
+        }
+
+        if (mExternalLogger.get() != null) {
```

Comment:
I remain very skeptical of `AtomicReference`'s usage here, please confirm that this should not be using `compareAndSet(null, newValue)`

Reply:
Thanks a lot for the comments. Let me do some research and think further about it.